File: PlaneBuilder/impl/Era.ts
/// <reference path="./Part.ts" />
/// <reference path="./Stats.ts" />

class Era extends Part {
    private vals: { name: string, maxbomb: number, cant_lift: number, stats: Stats }[];
    private selected: number;
    constructor(js: JSON) {
        super();
        this.selected = 0;
        this.vals = [];
        for (let elem of js["options"]) {
            var opt = {
                name: elem["name"], maxbomb: elem["maxbomb"],
                cant_lift: elem["cant_lift"], stats: new Stats(elem),
            };
            this.vals.push(opt);
        }
    }

    public toJSON() {
        return {
            selected: this.selected
        };
    }

    public fromJSON(js: JSON, json_version: number) {
        this.selected = js["selected"];
        if (json_version < 10.35) {
            if (this.selected > 2) {
                this.selected++;
            }
        }
    }

    public serialize(s: Serialize) {
        s.PushNum(this.selected);
    }

    public deserialize(d: Deserialize) {
        this.selected = d.GetNum();
    }

    public GetSelected() {
        return this.selected;
    }

    public GetSelectedText() {
        return this.vals[this.selected].name;
    }

    public SetSelected(index: number) {
        this.selected = index;
        this.CalculateStats();
    }

    public GetEraOptions(): { name: string }[] {
        return this.vals;
    }

    public GetLiftBleed(): number {
        return this.vals[this.selected].stats.liftbleed;
    }

    public GetMaxBomb(): number {
        return this.vals[this.selected].maxbomb;
    }

    public GetCantLift(): number {
        return this.vals[this.selected].cant_lift;
    }

    public PartStats(): Stats {
        var s = new Stats();
        s = s.Add(this.vals[this.selected].stats);
        return s;
    }

    public SetCalculateStats(callback: () => void) {
        this.CalculateStats = callback;
    }

    public GetElectrics(): { storage: number, equipment: { source: string, charge: string }[] } {
        let value = { storage: 0, equipment: [] };
        return value;
    }
}

File: flyingcircusrust/src/era.rs
use ui_core::*;
use ui_macro::*;

use crate::json::*;
use crate::part::{ElectricsMessage, Part};
use crate::serialization::{JSSerializable, Serializable};
use crate::stats::Stats;

pub struct EraEntry {
    pub name: String,
    pub maxbomb: f32,
    pub cant_lift: i32,
    pub stats: Stats,
}

#[derive(UIBindings)]
pub struct Era {
    vals: Vec<EraEntry>,
    #[ui(select, source = "vals")]
    pub selected_era: usize,
}

impl Era {
    pub fn new(vals: Vec<EraEntry>) -> Era {
        Era {
            vals,
            selected_era: 0,
        }
    }

    pub fn set_state(&mut self, era: usize) {
        self.selected_era = era;
    }

    pub fn get_cant_lift(&self) -> i16 {
        self.vals[self.selected_era].cant_lift as i16
    }

    /// Get maximum bomb load ratio for the currently selected era
    /// TypeScript: GetMaxBomb()
    /// Returns the maxbomb value (ratio) for calculating bomb capacity
    /// Values range from 0.1667 (Pioneer) to 0.5 (Last Hurrah)
    pub fn get_max_bomb(&self) -> f32 {
        self.vals[self.selected_era].maxbomb
    }

    /// Get the name of the currently selected era
    /// TypeScript: GetSelectedText()
    pub fn get_selected_text(&self) -> String {
        self.vals[self.selected_era].name.clone()
    }
}

impl Part for Era {
    fn part_stats(&mut self) -> Stats {
        self.vals[self.selected_era].stats.clone()
    }
    fn get_electrics(&self) -> crate::part::ElectricsMessage {
        ElectricsMessage::new()
    }
}

impl Serializable for Era {
    fn deserialize(
        &mut self,
        d: &mut crate::serialization::Deserializer,
    ) -> Result<(), crate::serialization::Error> {
        self.selected_era = d.get_num()? as usize;
        Ok(())
    }

    fn serialize(
        &self,
        s: &mut crate::serialization::Serializer,
    ) -> Result<(), crate::serialization::Error> {
        s.push_num(self.selected_era as i16)?;
        Ok(())
    }
}

impl JSSerializable for Era {
    fn to_json(&self) -> serde_json::Value {
        let mut map = serde_json::Map::new();
        map.insert("selected".to_string(), self.selected_era.into());

        serde_json::Value::Object(map)
    }

    fn from_json(&mut self, js: &serde_json::Value, json_version: f32) {
        self.selected_era = jsnum(js, "selected") as usize;
        if json_version < 10.35 {
            if self.selected_era > 2 {
                self.selected_era += 1;
            }
        }
    }
}
